Preserve scroll position in plugin settings tab

mnaoumov opened the pull request:
When `Hot Reload` reloads the plugin which has its settings tab open, the settings tab is closed and the scroll position in it disappears. This PR fixes that problem.

pjeby commented:
Hm.  This looks like it has a few potential failure modes, such as the possibility of changing tabs during reload, a late-registered setting tab, closing the tab during enable, multiple plugins updating, dynamically loaded settings, and so on.  At the very least it should confirm that another tab hasn't been selected (or the settings closed) while the enable was taking place, I think.

mnaoumov commented:
@pjeby Thanks for the quick review. I addressed your concerns.

> the possibility of changing tabs during reload

addressed in the new commit I made

> a late-registered setting tab

covered by my original commit in the [`null` check](https://github.com/pjeby/hot-reload/pull/27/commits/a4fd181c7206244007a624db6c1f9e2b806d8efd#diff-8866fe82788fa69551db122e042e86bd762938f9fcf6c91971473414edcc9283R133)

> closing the tab during enable

addressed in the new commit I made

> multiple plugins updating

Regardless of how many plugins you update, you can have not more than one settings tab opened before reload happens, so I don't see any probable failures in this scenario.

> dynamically loaded settings, and so on.

Well, there are potential scenarios of poorly written plugin settings tab, when opening it too early could cause data loss, which should not happen in normal scenarios.

If you want, I can add a setting tab for `Hot Reload` to opt-out the new behavior.

> it should confirm that another tab hasn't been selected

addressed in the new commit I made

> (or the settings closed) while the enable was taking place

addressed in the new commit I made

mnaoumov commented:
> If you want, I can add a setting tab for `Hot Reload` to opt-out the new behavior.

Actually did it in the new commit

pjeby commented:
I don't plan to add settings to the plugin at this time, and if I do later I want to use the Ophidian-generated settings UI, which all my plugins that do use settings are based on.  So I would want this feature compatible with those.  That means a possibly async-added settings tab, so a slightly more complex approach is required for this to work.  I'll take a look at it this weekend.

mnaoumov commented:
@pjeby I have an idea how to handle async added settings tab. I'll update the PR shortly

pjeby commented:
The thing I was going to try was to use around() on addSettingsTab in Plugin.prototype, something like:

```ts
async reload(plugin: string) {
    const plugins = this.app.plugins;

    // Don't reload disabled plugins
    if (!plugins.enabledPlugins.has(plugin)) return;

    let disposeTabHandler: () => void = undefined;

    if (this.app.setting.activeTab?.id === plugin) {
        const {scrollTop, scrollLeft} = this.app.setting.activeTab.containerEl;
        disposeTabHandler = around(Plugin.prototype, {
            addSettingTab(next) {
                return function(this: Plugin, ...args: any[]) {
                    if (this.manifest.id === plugin) {
                        disposeTabHandler()
                        // Let registration and any other microtasks finish first
                        setTimeout(()=> {
                            if (this.app.setting.containerEl.isShown() && this.app.setting.activeTab === null) {
                                this.app.setting.openTabById(plugin)?.containerEl.scrollTo({left: scrollLeft, top: scrollTop});
                            }
                        }, 0);
                    }
                    return next.apply(this, args);
                }
            },
        })
    }

    await plugins.disablePlugin(plugin);
    console.debug("disabled", plugin);

    // Ensure sourcemaps are loaded (Obsidian 0.14+)
    const oldDebug = localStorage.getItem("debug-plugin");
    localStorage.setItem("debug-plugin", "1");
    const uninstall = preventSourcemapStripping(this.app, plugin)
    try {
        await plugins.enablePlugin(plugin);
    } finally {
        // Restore previous setting
        if (oldDebug === null) localStorage.removeItem("debug-plugin"); else localStorage.setItem("debug-plugin", oldDebug);
        uninstall?.()
        if (disposeTabHandler) {
            // if the plugin was loaded, arrange for it to dispose the handler, else dispose it ourselves
            if (plugins.plugins[plugin]) plugins.plugins[plugin].register(disposeTabHandler); else disposeTabHandler();
        }
    }
    console.debug("enabled", plugin);
    new Notice(`Plugin "${plugin}" has been reloaded`);
}
```

But that's just an untested sketch at the moment.  It should correctly handle the case, I think, where a plugin does some async work before registering its settings tab, such that it's not immediately post-enable.

mnaoumov commented:
@pjeby yes, I had a similar idea in mind but simpler. I'll send my PR in a bit

mnaoumov commented:
@pjeby Added a simpler version

pjeby commented:
I went with a slightly different approach that doesn't do the monkeypatching until/unless it's actually used, to avoid inline cache purges during vault startup.  I also tweaked the logic a bit to prevent some race conditions that probably wouldn't have ever happened but still bugged me.  :wink:
